src/ykfde-enroll

```diff
@@ -103,67 +118,124 @@ printf '%s\n' "WARNING: This script will utilize LUKS keyslot '$LUKS_KEYSLOT' on
-    YKFDE_PASSPHRASE="$YKFDE_CHALLENGE$YKFDE_RESPONSE"
-  else
-    YKFDE_PASSPHRASE="$YKFDE_RESPONSE"
+  echo "   Remember to touch the old device if necessary."
+
+  if [ "$DBG" ]; then read -r YKFDE_OLD_PASSPHRASE; else read -r -s YKFDE_OLD_PASSPHRASE; fi
+
+  [ "$DBG" ] && printf '%s\n' "   Running: 'ykchalresp -$YKFDE_CHALLENGE_SLOT $YKFDE_OLD_CHALLENGE'..."
+  # we have to temporary disable '-e' shell option otherwise there would bo no chance for second attempt
+  set +e
+  YKFDE_OLD_RESPONSE="$(ykchalresp -"$YKFDE_CHALLENGE_SLOT" "$YKFDE_OLD_CHALLENGE" 2>/dev/null | tr -d '\n')"
+  set -e
+  if [ -z "$YKFDE_OLD_RESPONSE" ]; then
+    sleep 3
+    echo "   YubiKey did not provide a response - Initializing second attempt, touch the device if necessary."
+    YKFDE_OLD_RESPONSE="$(ykchalresp -"$YKFDE_CHALLENGE_SLOT" "$YKFDE_OLD_CHALLENGE" | tr -d '\n')"
   fi
-fi
 
-echo "Please provide the old LUKS passphrase for the existing keyslot."
-printf " Enter passphrase: "
-if [ "$DBG" ]; then read -r YKFDE_OLD_PASSPHRASE; else read -r -s YKFDE_OLD_PASSPHRASE; fi
+  [ "$DBG" ] && printf '%s\n' "   Received response: '$YKFDE_OLD_RESPONSE'"
+
+  if [ "$YKFDE_OLD_RESPONSE" ]; then
+    if [ "$YKFDE_CHALLENGE_PASSWORD_NEEDED" ]; then
+      YKFDE_OLD_PASSPHRASE="$YKFDE_OLD_CHALLENGE$YKFDE_OLD_RESPONSE"
+    else
+      YKFDE_OLD_PASSPHRASE="$YKFDE_RESPONSE"
+    fi
+  fi
+else
+  echo "Please provide the old LUKS passphrase for the existing keyslot."
+  printf " Enter passphrase: "
+  if [ "$DBG" ]; then read -r YKFDE_OLD_PASSPHRASE; else read -r -s YKFDE_OLD_PASSPHRASE; fi
+fi
 
 if [ "$YKFDE_PASSPHRASE" ]; then
   [ "$DBG" ] && printf '%s\n' " > Passing '$YKFDE_PASSPHRASE' to 'cryptsetup'"
   if [ "$CHANGE_KEYSLOT" ]; then
     [ "$DBG" ] && printf '%s\n' " > Changing LUKS passphrase with 'cryptsetup --key-slot=$LUKS_KEYSLOT luksChangeKey $YKFDE_LUKS_DEV'..." || echo " > Changing LUKS passphrase with 'cryptsetup'..."
     printf '%s\n' "$YKFDE_OLD_PASSPHRASE" "$YKFDE_PASSPHRASE" "$YKFDE_PASSPHRASE" | cryptsetup --key-slot="$LUKS_KEYSLOT" luksChangeKey "$YKFDE_LUKS_DEV" 2>&1
+    echo "   LUKS passphrase for key slot $LUKS_KEYSLOT successfully changed"
   else
     [ "$DBG" ] && printf '%s\n' " > Adding new LUKS passphrase with 'cryptsetup --key-slot=$LUKS_KEYSLOT luksAddKey $YKFDE_LUKS_DEV'..." || echo " > Adding new LUKS passphrase with 'cryptsetup'..."
     printf '%s\n' "$YKFDE_OLD_PASSPHRASE" "$YKFDE_PASSPHRASE" "$YKFDE_PASSPHRASE" | cryptsetup --key-slot="$LUKS_KEYSLOT" luksAddKey "$YKFDE_LUKS_DEV" 2>&1
+    echo "   New LUKS passphrase successfully added"
   fi
+elif [ "$YKFDE_SLOT_KILL" ]; then
+  [ "$DBG" ] && printf '%s\n' " > Killing with 'cryptsetup luksKillSlot $YKFDE_LUKS_DEV $LUKS_KEYSLOT'..." || echo " > Killing slot 'cryptsetup'..."
+  printf %s "$YKFDE_OLD_PASSPHRASE" | cryptsetup luksKillSlot "$YKFDE_LUKS_DEV" "$LUKS_KEYSLOT" 2>&1
+  echo "   LUKS key slot $LUKS_KEYSLOT successfully killed"
```

Comment:
If a string to be printed has `$variable` please use `printf '%s\n'` instead of `echo`.